api/src/main/java/com/velocitypowered/api/event/connection/PreLoginEvent.java

```diff
@@ -0,0 +1,50 @@
+package com.velocitypowered.api.event.connection;
+
+import com.google.common.base.Preconditions;
+import com.velocitypowered.api.event.ResultedEvent;
+import com.velocitypowered.api.proxy.InboundConnection;
+
+import javax.annotation.Nonnull;
+
+/**
+ * This event is fired when a player has initiated a connection with the proxy but before the proxy authenticates the
+ * player with Mojang or before the player's proxy connection is fully established (for offline mode).
+ */
+public class PreLoginEvent implements ResultedEvent<ResultedEvent.ComponentResult> {
+    private final InboundConnection connection;
+    private final String username;
+    private ComponentResult result;
+
+    public PreLoginEvent(InboundConnection connection, String username) {
```

Comment:
These should also be `@NonNull` and have the not null check too.

Reply:
I am waiting on adding in full checker framework support.